File: 015.NNMultipleDelegate/NNMultipleDelegate/ViewController.m
//
//  ViewController.m
//  NNMultipleDelegate
//
//  Created by 刘朋坤 on 2018/8/17.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "ViewController.h"
#import "NNMultipleDelegateHelper.h"
#import "NNStretchHeaderView.h"
#import "NNLabel.h"

@interface ViewController ()<UITableViewDelegate, UITableViewDataSource>

/** UITableView */
@property (nonatomic, strong) UITableView *tableView;
/** 头部视图 */
@property (nonatomic, strong) NNStretchHeaderView *headerView;
/** 测试偏移量 */
@property (nonatomic, strong) NNLabel *contentOffsetLabel;
@property (nonatomic, strong) NNMultipleDelegateHelper *delegateHelper;

@end

@implementation ViewController
static NSString *const UITableViewCellID = @"UITableViewCellID";

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.headerView];
    [self.view addSubview:self.tableView];
    [self.view addSubview:self.contentOffsetLabel];
    [self creatDelegateHelper];
}

- (void)creatDelegateHelper {
    _delegateHelper = [NNMultipleDelegateHelper new];
    _delegateHelper.delegateArray = @[self, self.headerView, self.contentOffsetLabel];
    self.tableView.dataSource = (id<UITableViewDataSource>)_delegateHelper;
    self.tableView.delegate = (id<UITableViewDelegate>)_delegateHelper;
}

#pragma mark - 代理区域
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UITableViewCellID forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"%ld", indexPath.row];
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 100;
}

#pragma mark - 懒加载区域
- (NNStretchHeaderView *)headerView {
    if (!_headerView) {
        _headerView = [[NNStretchHeaderView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 200)];
    }
    return _headerView;
}

- (UITableView *)tableView {
    if (!_tableView) {
        _tableView = [[UITableView alloc] initWithFrame:self.view.frame style:UITableViewStylePlain];
        _tableView.dataSource = self;
        _tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
        _tableView.contentInset = UIEdgeInsetsMake(200, 0, 0, 0);
        _tableView.backgroundColor = [UIColor clearColor];
        _tableView.delegate = self;
        [_tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:UITableViewCellID];
    }
    return _tableView;
}

- (NNLabel *)contentOffsetLabel {
    if (!_contentOffsetLabel) {
        _contentOffsetLabel = [[NNLabel alloc] initWithFrame:CGRectMake(self.view.frame.size.width / 2.0 - 50, 300, 100, 21)];
    }
    return _contentOffsetLabel;
}

@end

File: 017.NN_UIViewAnimationPractice_swift/NN_UIViewAnimationPractice_swift/ViewController.swift
//
//  ViewController.swift
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    let titleArray = ["Frame Bounds Center", "Transform", "Transition", "SpringAnimation", "Alpha", "BackgroundColor"]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(tableView)
    }

    // 懒加载
    lazy var tableView : UITableView = {
        let tempTableView = UITableView(frame: view.frame, style: .plain)
        tempTableView.delegate = self
        tempTableView.dataSource = self
        tempTableView.tableFooterView = UIView()
        return tempTableView
    }()
    
}

/** 代理区域 */
extension ViewController:UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return titleArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellID = "cellID"
        var cell = tableView.dequeueReusableCell(withIdentifier: cellID)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: cellID)
        }
        cell?.textLabel?.text = titleArray[indexPath.row]
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        switch indexPath.row {
        case 0:
            self.navigationController?.pushViewController(NNFrameAnimationController(), animated: true)
            break
        case 1:
            self.navigationController?.pushViewController(NNTransformController(), animated: true)
            break
        case 2:
            self.navigationController?.pushViewController(NNTransitionAnimationController(), animated: true)
            break
        case 3:
            self.navigationController?.pushViewController(NNSpringAnimationController(), animated: true)
            break
        case 4:
            self.navigationController?.pushViewController(NNAlphaAnimationViewController(), animated: true)
            break
        case 5:
            self.navigationController?.pushViewController(NNBackgroundAnimationViewController(), animated: true)
            break
        default: break
        }
    }
}

